src/routes/bounce_rules_activity.jsx

```diff
@@ -0,0 +1,228 @@
+      });
+    }
+  }
+
+  logout() {
+    const { history } = this.props;
+    localStorage.clear();
+    history.push("/");
+  }
+
+  handleModalClose(e) {
+    const { id } = e.currentTarget;
+    this.setState({
+      [id]: false,
+      isInvalidInput: false,
+      selectedRule: {},
+    });
+  }
+
+  updateSearchToken(e) {
+    this.setState({
+      searchToken: e.target.value.toLowerCase(),
+    });
+  }
+
+  updateSearchCategory(e) {
+    this.setState({
+      searchCategory: e.value.toLowerCase(),
+    });
+  }
+
+  filterRules(rules) {
+    const { searchToken } = this.state;
+    return rules.filter(
+      rule =>
+        rule.bounce_action.toLowerCase().includes(searchToken) ||
+        rule.description.toLowerCase().includes(searchToken)
+    );
+  }
+
+  paginate(activityLog) {
+    const { currentActivityPageIndex, rulesToShow } = this.state;
+    const ruleStartIndex = (currentActivityPageIndex - 1) * rulesToShow;
+    const ruleEndIndex =
+      (currentActivityPageIndex - 1 * currentActivityPageIndex + rulesToShow) *
+      currentActivityPageIndex;
+    return activityLog.slice(ruleStartIndex, ruleEndIndex);
+  }
+
+  updatePageIndex(e) {
+    const newIndex = parseInt(e.currentTarget.getAttribute("value"), 10);
+    this.setState(prevState => {
+      const isPageIndexUpdated =
+        prevState.currentActivityPageIndex !== newIndex;
+      return {
+        currentActivityPageIndex: isPageIndexUpdated
+          ? newIndex
+          : prevState.currentActivityPageIndex,
+      };
+    });
+  }
+
+  handlePrevClicked() {
+    this.setState(prevState => ({
+      currentActivityPageIndex: prevState.currentActivityPageIndex - 1,
+    }));
+  }
+
+  handleNextClicked() {
+    this.setState(prevState => ({
+      currentActivityPageIndex: prevState.currentActivityPageIndex + 1,
+    }));
+  }
+
+  isDuplicate(searchCategory, searchToken) {
+    let found = false;
+    const { filterOptions } = this.state;
+    filterOptions.filter(filter => {
+      if (
+        filter.searchCategory === searchCategory &&
+        filter.searchToken === searchToken
+      ) {
+        found = true;
+      }
+      return false;
+    });
+    return found;
+  }
+
+  addFilter() {
+    const { searchCategory, searchToken } = this.state;
+    if (!searchCategory || !searchToken) {
+      this.setState({
+        invalidFilter: true,
+      });
+      return;
+    }
+    if (!this.isDuplicate(searchCategory, searchToken)) {
+      this.setState(prevState => ({
+        invalidFilter: false,
+        filterOptions: [
+          ...prevState.filterOptions,
+          { searchCategory, searchToken },
+        ],
+        searchToken: "",
+      }));
+    } else {
+      this.setState({
+        invalidFilter: true,
+      });
+    }
+  }
+
+  removeFilter(filter) {
+    const { filterOptions } = this.state;
+    const newFilter = [...filterOptions];
+    const index = newFilter.indexOf(filter);
+    newFilter.splice(index, 1);
```

Comment:
Isn't a filterOption an object so doing an indexOf the object inside an array of objects won't always match the way we expect if it's not a string/primitive value? Let's also avoid using splice since it's not functional and edits the existing array. Let's continue to use the functional array approaches cause JS have so many great array functional helpers which always return a new array and don't modify the existing array. we can do something like this to check for objects that aren't the same ```const newFilterOptions = filterOptions.filter(filterOption => filterOption.searchCategory !== filter.searchCategory && filterOption.searchToken !== filterOption.searchToken)```